File: Cpp_Version/src/distance_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "turtlesim/msg/pose.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "std_msgs/msg/bool.hpp"
#include "turtlesim/srv/set_pen.hpp"
#include "turtlesim/srv/teleport_absolute.hpp"

#include <cmath>
#include <chrono>
#include <memory>
#include <string>
#include <functional>

// -------------------------
// Standard Library aliases
// -------------------------
using std::string;
using std::bind;
using std::placeholders::_1;
using std::sqrt;
using std::chrono::milliseconds;


// -------------------------
// ROS2 Type Aliases
// -------------------------
using rclcpp::Node;
using rclcpp::sleep_for;
using rclcpp::init;
using rclcpp::spin;
using rclcpp::shutdown;

// -------------------------
// Generic Aliases
// -------------------------
template<typename T>
using Sub = rclcpp::Subscription<T>;

template<typename T>
using Pub = rclcpp::Publisher<T>;

using Timer = rclcpp::TimerBase;

using Pose           = turtlesim::msg::Pose;
using Twist          = geometry_msgs::msg::Twist;
using BoolMsg        = std_msgs::msg::Bool;

using SetPenSrv      = turtlesim::srv::SetPen;
using TeleportSrv    = turtlesim::srv::TeleportAbsolute;

using ClientSetPen   = rclcpp::Client<SetPenSrv>;
using ClientTeleport = rclcpp::Client<TeleportSrv>;

using SetPenReq      = SetPenSrv::Request;
using TeleportReq    = TeleportSrv::Request;

// -------------------------
// ANSI Colors
// -------------------------
static const string Y = "\033[33m";
static const string R = "\033[0m";


// =====================================================
// CLASS NODE
// =====================================================
class DistanceNode : public Node
{
public:
    DistanceNode() : Node("distance_node")
    {
        RCLCPP_INFO(
            this->get_logger(),
            "%sdistance_node v4.3 loaded%s",
            Y.c_str(), R.c_str()
        );

        // -------------------------
        // Subscribers
        // -------------------------
        sub1_ = create_subscription<Pose>(
            "/turtle1/pose", 10,
            bind(&DistanceNode::pose1Callback, this, _1)
        );

        sub2_ = create_subscription<Pose>(
            "/turtle2/pose", 10,
            bind(&DistanceNode::pose2Callback, this, _1)
        );

        // -------------------------
        // Publishers
        // -------------------------
        pub1_ = create_publisher<Twist>("/turtle1/cmd_vel", 10);
        pub2_ = create_publisher<Twist>("/turtle2/cmd_vel", 10);

        freeze_pub_ = create_publisher<BoolMsg>("/freeze_turtles", 10);

        // -------------------------
        // Service Clients
        // -------------------------
        pen1_client_ = create_client<SetPenSrv>("/turtle1/set_pen");
        pen2_client_ = create_client<SetPenSrv>("/turtle2/set_pen");
        tp1_client_  = create_client<TeleportSrv>("/turtle1/teleport_absolute");
        tp2_client_  = create_client<TeleportSrv>("/turtle2/teleport_absolute");

        // -------------------------
        // Timer
        // -------------------------
        timer_ = create_wall_timer(
            milliseconds(50),
            bind(&DistanceNode::update, this)
        );
    }

private:

    // =====================================================
    // FREEZE CONTROL
    // =====================================================
    void setFreeze(bool state)
    {
        BoolMsg msg;
        msg.data = state;
        freeze_pub_->publish(msg);
    }


    // =====================================================
    // UTILITIES
    // =====================================================
    void setPen(const ClientSetPen::SharedPtr &client,
                int r, int g, int b,
                int width, bool off)
    {
        if (!client->wait_for_service(milliseconds(100)))
            return;

        auto req = std::make_shared<SetPenReq>();
        req->r = r;
        req->g = g;
        req->b = b;
        req->width = width;
        req->off = off ? 1 : 0;

        client->async_send_request(req);
    }

    void teleport(const ClientTeleport::SharedPtr &client,
                  const Pose &p)
    {
        if (!client->wait_for_service(milliseconds(100)))
            return;

        auto req = std::make_shared<TeleportReq>();
        req->x = p.x;
        req->y = p.y;
        req->theta = p.theta;

        client->async_send_request(req);
    }

    void stopTurtle(const Pub<Twist>::SharedPtr &pub)
    {
        Twist msg;
        msg.linear.x = 0;
        msg.angular.z = 0;
        pub->publish(msg);
    }

    bool isNearWall(const Pose &p)
    {
        return (p.x < 1.5 || p.x > 9.5 ||
                p.y < 1.5 || p.y > 9.5);
    }


    // =====================================================
    // CALLBACKS
    // =====================================================
    void pose1Callback(const Pose &p)
    {
        pose1_ = p;
        pose1_ready_ = true;

        if (!init1_) {
            last_stop1_ = p;
            init1_ = true;
            setPen(pen1_client_, 0, 0, 255, 3, false);
        }
    }

    void pose2Callback(const Pose &p)
    {
        pose2_ = p;
        pose2_ready_ = true;

        if (!init2_) {
            last_stop2_ = p;
            init2_ = true;
            setPen(pen2_client_, 0, 0, 255, 3, false);
        }
    }


    // =====================================================
    // MAIN LOOP
    // =====================================================
    void update()
    {
        if (!pose1_ready_ || !pose2_ready_)
            return;

        // Save stop positions
        if (pose1_.linear_velocity == 0 && pose1_.angular_velocity == 0)
            last_stop1_ = pose1_;

        if (pose2_.linear_velocity == 0 && pose2_.angular_velocity == 0)
            last_stop2_ = pose2_;


        // -------------------------
        // Turtle–turtle collision
        // -------------------------
        float dx = pose1_.x - pose2_.x;
        float dy = pose1_.y - pose2_.y;

        if (sqrt(dx * dx + dy * dy) < 1.5)
        {
            RCLCPP_ERROR(this->get_logger(), "[COLLISION] turtles");

            setFreeze(true);
            stopTurtle(pub1_);
            stopTurtle(pub2_);

            setPen(pen1_client_, 255, 0, 0, 3, false);
            setPen(pen2_client_, 255, 0, 0, 3, false);

            sleep_for(milliseconds(400));

            setPen(pen1_client_, 0, 0, 0, 3, true);
            setPen(pen2_client_, 0, 0, 0, 3, true);

            teleport(tp1_client_, last_stop1_);
            teleport(tp2_client_, last_stop2_);

            stopTurtle(pub1_);
            stopTurtle(pub2_);

            setPen(pen1_client_, 0, 0, 255, 3, false);
            setPen(pen2_client_, 0, 0, 255, 3, false);

            stopTurtle(pub1_);
            stopTurtle(pub2_);

            setFreeze(false);
            return;
        }


        // -------------------------
        // Wall collision (turtle1)
        // -------------------------
        if (isNearWall(pose1_))
        {
            RCLCPP_ERROR(this->get_logger(), "[COLLISION] turtle1 wall");

            setFreeze(true);
            stopTurtle(pub1_);

            setPen(pen1_client_, 255, 0, 0, 3, false);
            sleep_for(milliseconds(400));

            setPen(pen1_client_, 0, 0, 0, 3, true);
            teleport(tp1_client_, last_stop1_);

            stopTurtle(pub1_);
            setPen(pen1_client_, 0, 0, 255, 3, false);
            stopTurtle(pub1_);

            setFreeze(false);
            return;
        }


        // -------------------------
        // Wall collision (turtle2)
        // -------------------------
        if (isNearWall(pose2_))
        {
            RCLCPP_ERROR(this->get_logger(), "[COLLISION] turtle2 wall");

            setFreeze(true);
            stopTurtle(pub2_);

            setPen(pen2_client_, 255, 0, 0, 3, false);
            sleep_for(milliseconds(400));

            setPen(pen2_client_, 0, 0, 0, 3, true);
            teleport(tp2_client_, last_stop2_);

            stopTurtle(pub2_);
            setPen(pen2_client_, 0, 0, 255, 3, false);
            stopTurtle(pub2_);

            setFreeze(false);
            return;
        }
    }

    // =====================================================
    // MEMBER VARIABLES
    // =====================================================
    Pose pose1_, pose2_;
    Pose last_stop1_, last_stop2_;
    
    bool pose1_ready_ = false;
    bool pose2_ready_ = false;
    bool init1_ = false;
    bool init2_ = false;
    
    // Subscribers
    Sub<Pose>::SharedPtr sub1_;
    Sub<Pose>::SharedPtr sub2_;
    
    // Publishers
    Pub<Twist>::SharedPtr    pub1_;
    Pub<Twist>::SharedPtr    pub2_;
    Pub<BoolMsg>::SharedPtr  freeze_pub_;
    
    // Service clients
    ClientSetPen::SharedPtr    pen1_client_;
    ClientSetPen::SharedPtr    pen2_client_;
    ClientTeleport::SharedPtr  tp1_client_;
    ClientTeleport::SharedPtr  tp2_client_;
    
    // Timer
    Timer::SharedPtr timer_;
};

// =====================================================
// MAIN
// =====================================================
int main(int argc, char **argv)
{
    init(argc, argv);
    spin(std::make_shared<DistanceNode>());
    shutdown();
    return 0;
}

File: assignment1_rt_py/assignment1_rt_py/distance_node.py
#!/usr/bin/env python3

import math
import time

import rclpy
from rclpy.node import Node

from turtlesim.msg import Pose
from geometry_msgs.msg import Twist
from std_msgs.msg import Bool

from turtlesim.srv import SetPen, TeleportAbsolute

# ANSI colors
Y = "\033[33m"
R = "\033[0m"

# ================================
# CONFIG
# ================================
COLLISION_DISTANCE = 1.5
WALL_THRESHOLD = 1.5
SLEEP_AFTER_COLLISION = 0.4  # seconds


class DistanceNode(Node):

    def __init__(self):
        super().__init__("distance_node")

        self.get_logger().info(f"{Y}distance_node 1.0 loaded{Y}")

        # -----------------------------
        # Internal state
        # -----------------------------
        self.pose1 = None
        self.pose2 = None

        self.last_stop1 = None
        self.last_stop2 = None

        self.init1 = False
        self.init2 = False

        # -----------------------------
        # Subscribers
        # -----------------------------
        self.create_subscription(Pose, "/turtle1/pose",
                                 self.pose1_callback, 10)
        self.create_subscription(Pose, "/turtle2/pose",
                                 self.pose2_callback, 10)

        # -----------------------------
        # Publishers
        # -----------------------------
        self.pub1 = self.create_publisher(Twist, "/turtle1/cmd_vel", 10)
        self.pub2 = self.create_publisher(Twist, "/turtle2/cmd_vel", 10)
        self.freeze_pub = self.create_publisher(Bool, "/freeze_turtles", 10)

        # -----------------------------
        # Service Clients
        # -----------------------------
        self.pen1 = self.create_client(SetPen, "/turtle1/set_pen")
        self.pen2 = self.create_client(SetPen, "/turtle2/set_pen")
        self.tp1 = self.create_client(TeleportAbsolute,
                                      "/turtle1/teleport_absolute")
        self.tp2 = self.create_client(TeleportAbsolute,
                                      "/turtle2/teleport_absolute")

        # -----------------------------
        # Timer
        # -----------------------------
        self.create_timer(0.05, self.update)

    # =====================================================
    # UTILITY FUNCTIONS
    # =====================================================

    def freeze(self, state: bool):
        msg = Bool()
        msg.data = state
        self.freeze_pub.publish(msg)

    def stop_turtle(self, pub):
        msg = Twist()
        pub.publish(msg)

    def set_pen(self, client, r, g, b, w, off):
        if not client.wait_for_service(timeout_sec=0.1):
            return

        req = SetPen.Request()
        req.r = r
        req.g = g
        req.b = b
        req.width = w
        req.off = 1 if off else 0
        client.call_async(req)

    def teleport(self, client, pose):
        if not client.wait_for_service(timeout_sec=0.1):
            return

        req = TeleportAbsolute.Request()
        req.x = pose.x
        req.y = pose.y
        req.theta = pose.theta
        client.call_async(req)

    @staticmethod
    def near_wall(p):
        return (
            p.x < WALL_THRESHOLD
            or p.x > 10.0 - WALL_THRESHOLD
            or p.y < WALL_THRESHOLD
            or p.y > 10.0 - WALL_THRESHOLD
        )

    # =====================================================
    # SUBSCRIBER CALLBACKS
    # =====================================================

    def pose1_callback(self, msg: Pose):
        self.pose1 = msg
        if not self.init1:
            self.init1 = True
            self.last_stop1 = msg
            self.set_pen(self.pen1, 0, 0, 255, 3, False)

        if msg.linear_velocity == 0 and msg.angular_velocity == 0:
            self.last_stop1 = msg

    def pose2_callback(self, msg: Pose):
        self.pose2 = msg
        if not self.init2:
            self.init2 = True
            self.last_stop2 = msg
            self.set_pen(self.pen2, 0, 0, 255, 3, False)

        if msg.linear_velocity == 0 and msg.angular_velocity == 0:
            self.last_stop2 = msg

    # =====================================================
    # MAIN LOOP
    # =====================================================

    def update(self):
        if self.pose1 is None or self.pose2 is None:
            return

        # --------------------------------
        # TURTLE–TURTLE COLLISION
        # --------------------------------
        dx = self.pose1.x - self.pose2.x
        dy = self.pose1.y - self.pose2.y
        dist = math.sqrt(dx * dx + dy * dy)

        if dist < COLLISION_DISTANCE:
            self.handle_collision("turtles", both=True)
            return

        # --------------------------------
        # WALL COLLISIONS
        # --------------------------------
        if self.near_wall(self.pose1):
            self.handle_collision("turtle1 wall", t1=True)

        if self.near_wall(self.pose2):
            self.handle_collision("turtle2 wall", t2=True)

    # =====================================================
    # COLLISION HANDLING
    # =====================================================

    def handle_collision(self, msg, both=False, t1=False, t2=False):
        self.get_logger().error(f"[COLLISION] {msg}")

        self.freeze(True)

        # Stop turtles
        if both or t1:
            self.stop_turtle(self.pub1)
        if both or t2:
            self.stop_turtle(self.pub2)

        # Flash red
        if both or t1:
            self.set_pen(self.pen1, 255, 0, 0, 3, False)
        if both or t2:
            self.set_pen(self.pen2, 255, 0, 0, 3, False)

        time.sleep(SLEEP_AFTER_COLLISION)

        # Disable pen
        if both or t1:
            self.set_pen(self.pen1, 0, 0, 0, 3, True)
        if both or t2:
            self.set_pen(self.pen2, 0, 0, 0, 3, True)

        # Teleport
        if both or t1:
            self.teleport(self.tp1, self.last_stop1)
        if both or t2:
            self.teleport(self.tp2, self.last_stop2)

        time.sleep(0.05)

        # Re-enable blue pen
        if both or t1:
            self.set_pen(self.pen1, 0, 0, 255, 3, False)
        if both or t2:
            self.set_pen(self.pen2, 0, 0, 255, 3, False)

        # Stop again to avoid drift
        if both or t1:
            self.stop_turtle(self.pub1)
        if both or t2:
            self.stop_turtle(self.pub2)

        self.freeze(False)


# =====================================================
# MAIN
# =====================================================
def main(args=None):
    rclpy.init(args=args)
    node = DistanceNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()
